python/ray/serve/api.py

```diff
@@ -433,14 +488,23 @@ def get_handle(self,
 
         if asyncio.get_event_loop().is_running() and sync:
             logger.warning(
-                "You are retrieving a ServeHandle inside an asyncio loop. "
+                "You are retrieving a sync ServeHandle inside an asyncio loop. "
                 "Try getting client.get_handle(.., sync=False) to get better "
                 "performance.")
-
-        if endpoint_name not in self._handle_cache:
-            handle = RayServeHandle(self._controller, endpoint_name, sync=sync)
-            self._handle_cache[endpoint_name] = handle
-        return self._handle_cache[endpoint_name]
+        if not asyncio.get_event_loop().is_running() and not sync:
+            logger.warning(
+                "You are retrieving an async ServHandle outside an asyncio loop. "
+                "You should make sure client.get_handle is called inside a running loop. "
+                "Or call client.get_handle(.., sync=True) to create synchrounous "
+                "handle.")
```

Comment:
I think we shouldn't do it automatically because user can use the sync handle even in async context for good reasons! To use async handle you have to modify all your code to be async compatible (move async everywhere).  additionally, a lot of code can run inside asyncio loop without users knowing it (e.g. jupyter, ipython)